proposals/0057-events.md

```diff
@@ -0,0 +1,261 @@
+buffer otherwise you will experience truncation fetching logs older than the
+RPCs local history, even when you configured your own RPC with a large log
+buffer. Another issue is that these logs are used to also record runtime
+information (compute units consumed) and plaintext debug messages, hence become
+very difficult to parse, which is why many developers avoid structured event
+logging and just use formatted strings.
+
+A lot of developers would prefer a reliable and universally compatible solution
+so recently Anchor has added a new feature called cpi_events. This is a
+convention on how to call your own program via CPI and encode a binary string
+as instruction data. It can be indexed via the RPC Transaction response, as
+innerInstruction data is always persisted. It is more reliable than sol_log_
+but it effectively reduces runtime limits like maximum CPI call depth for the
+actual application. It also causeses unnecessary overhead in replay as a new
+execution context neeeds to be created, for something as trivial as a log
+message. For the average program (1-2mb) that would be around 5000-9000 CU per
+1kb log message.
+
+In comparison to ethereum and cosmos, solana's preferred access for transaction
+meta-data is via address filtering, e.g. `getSignaturesForAddress` call.
+
+1. Topics: the ethereum yellow paper instead specifies that every event can be
+attributed to up to 4 topics. Each topic is a 32 byte subscription key, users
+can use to filter these events via RPC.
+
+This seems restrictive compared to the expressivenes that is guaranteed through
+the context of each instruction, program id and account metas, this data is
+anyways available at the RPC node.
+
+2. Schema: the cosmos sdk requires every event to have a protobuf schema, this
+makes it very easy for third parties to come in and build generalized indexers
+for the cosmos ecosystem. It also allows to specify complex filter queries that
+can evaluated on the RPC node to stream or retrieve events.
+
+Parsing the event payload is not a task for the rpc layer, and should instead
+be handled by 3rd party providers. As schemas on solana are flexible and can
+change with program updates, this would require access to historic schema
+definitions / account state.
+
+## New Terminology
+
+Discriminator - a unique n byte sequence often a 8 byte hash of a method
+signature or data type, that should be used to tag the type of a data structure
+and encoded always in the very first bytes of an event, account or instruction
+for the sake of efficient parsing.
+
+POA - Program Owned Account
+
+EOA - Externally Owned Account
+
+## Detailed Design
+
+### Runtime 
+
+In order to log an event, the SVM runtime must implement a new syscall
+`sol_emit_`. The runtime must charge 100 + ceil(48 * len / 1024) CU per
+invocation.
+
+```
+#include <sol/emit.h>
+void sol_emit_(/* r1 */ uint64_t len,
+               /* r2 */ const char *message);
```

Comment:
why? you can cpi with 5k instruction data right now for 1020 CU

Reply:
sorry, 5k was an example. i just think we need some reasonable event payload limit defined

Reply:
@ripatel-fd brought up in the discussion today that we should consider the cost of including a hash of all events in consensus. this would allow light-clients to verify the authenticity of events received from third party rpc providers. 

bank deltas are currently hashed via sha256, the sha256 syscall charges 85 + 0.5 CU / byte. Given that sha256 is kinda overkill for what we are trying to achieve a faster hash function would be adequate, e.g. blake3. Albeit blake3 being roughly 2x faster to compute (see [blake3-avx2 vs sha256-ni](https://github.com/BLAKE3-team/BLAKE3/issues/207)) the runtime currently charges the same amount for blake3 hashing.

Happy to go either way from here, but 100 + 0.25 * len CU would be a good approximation in this case. CPI for event logging would be only "cheaper" when using  >4k event payloads in this case, most events are <512b in length.
Limit per 200k CU: 800kB
Limit per 48M CU: 200MB

Reply:
@mvines please let me know, if this issue is resolved